### PR TITLE
test(exec_command): regression suite for output volume and filter correctness

### LINKED ISSUE #960: test(exec_command): regression suite for output volume and filter correctness
## Summary

Add a regression test suite that ensures `exec_command` output volume never silently regresses. Issues #958 (byte-cap) and #959 (output filters) introduce new truncation and filter paths with no existing test coverage. Without explicit tests, a refactor or filter-table change could reintroduce 100k-char responses without any CI signal. The suite also adds a `chars_threshold_breach` metric field so production regressions are observable without waiting for a session post-mortem.

## Context

- Issues #958 and #959 add new code paths (byte-cap, filter application) with no current test coverage in `tests/` or `crates/aptu-coder/src/lib.rs`
- Existing exec tests cover happy-path execution and timeout; they do not cover output size or filtering
- Session `20260523_52` showed 102,582 chars returned for a single `git pull` -- detectable only via manual session inspection. There is no automated signal in CI or metrics that would have flagged this before it happened
- `output_chars` is already recorded in every `MetricEvent`. A `chars_threshold_breach` boolean (fires when `output_chars > 30_000`, matching `MAX_STDOUT_BYTES` from #958) would make large responses visible in the JSONL metrics without adding log noise

Depends on: #958, #959 (tests validate behavior introduced in those issues).

## Data Basis for `chars_threshold_breach` Threshold

The threshold is set to 30,000 chars to match `MAX_STDOUT_BYTES` as established in #958. This value was derived from 27,981 observed `exec_command` calls:

| Percentile | output_chars |
|---|---|
| p95 | 7,104 |
| p99 | 17,479 |
| p99.5 | 24,414 |
| p99.7 | ~30,000 |
| p99.9 | 67,586 |
| max | 957,511 |

At 30k, `chars_threshold_breach` would fire for the top 0.33% of calls (91 calls), which account for a disproportionate share of total output volume. The original issue text proposed 20k; that is revised to 30k to stay consistent with the byte-cap constant in #958 and avoid false positives in the 20k-30k band (533 additional calls that complete successfully without truncation). The distinction:

- `output_truncated = true`: the byte cap or line cap fired; content was cut
- `chars_threshold_breach = true` (at 30k threshold): a large response reached the LLM without truncation, in the 20k-30k band; this is the early-warning signal for calls approaching the cap

The jq query to monitor this in production:
```
jq 'select(.chars_threshold_breach==true) | [.tool, .output_chars, (.session_id//"?")] | @tsv' metrics-*.jsonl | sort -t$'\t' -k2 -rn
```

## Prerequisites

- Depends on: #958
- Depends on: #959

### Strategy

1. **Unit tests for `handle_output_persist` byte-cap path (from #958)**

   File: `crates/aptu-coder/src/lib.rs` (inline test module, consistent with existing test pattern)

   Note: #958 already added 9 tests for the byte-cap paths. The following tests from the original plan are already covered and must not be duplicated:

   | Test | Covered by #958 |
   |---|---|
   | byte cap fires before line cap | `test_exec_line_and_byte_interaction` |
   | no cap under both limits | `test_exec_no_truncation_under_limits` |
   | combined cap on output_text | `test_exec_byte_overflow_combined_exceeds_50k` |
   | stderr byte cap independent | `test_exec_byte_overflow_stderr_exceeds_10k` |

   **One gap remains** (not covered by #958): the line-cap-fires-before-byte-cap path. Add only this test:

   - `test_line_cap_fires_before_byte_cap`: stdout with 2,500 short lines (each 5 chars, total ~12,500 chars -- well under 30k) -- line cap fires at 2,000 lines, byte cap does not fire, `output_truncated = true`, returned content is the last 50 lines

2. **Unit tests for filter application (from #959)**

   File: `crates/aptu-coder/src/lib.rs` or `crates/aptu-coder-core/src/` (wherever `FilterRule` and `apply_filter` are implemented)

   - `test_git_pull_filter_success`: simulate `git pull` stdout with 948 stat lines + summary line; assert output ≤ 200 chars and contains file/insertion/deletion counts
   - `test_git_pull_filter_no_op`: simulate `"Already up to date."` output; assert result is `"ok (up-to-date)"` or equivalent (≤ 30 chars)
   - `test_filter_not_applied_on_failure`: non-zero exit code; assert raw output is returned unchanged, filter not applied
   - `test_no_stat_injection`: command `"git pull origin main"` -- assert rewritten command contains `--no-stat`; command `"git pull --stat"` -- assert `--no-stat` NOT injected
   - `test_keep_lines_matching`: filter with `keep_lines_matching = ["^error"]`; assert only error lines survive
   - `test_on_empty_substitution`: filter with `on_empty = "ok"`, all lines stripped; assert output is `"ok"`
   - `test_project_local_overrides_builtin`: load a `FilterRule` via the TOML path for a command that also matches a built-in rule; assert project-local rule wins (first-match semantics)
   - `test_invalid_toml_falls_back_gracefully`: malformed `.aptu/filters.toml`; assert server starts, built-in rules apply, no panic

3. **Integration smoke test: `git pull` volume guard**

   File: `tests/exec_volume.rs` (integration test, requires a bare git repo fixture)

   - Create a temporary git repo with ≥ 50 commits and ≥ 200 changed files (use a script to generate synthetic history)
   - Run `exec_command` with `"git pull"` against it
   - Assert: `output.interleaved.len() < 2_000` (chars, not lines)
   - Do NOT assert `output_truncated == false`. The byte cap (#958) is the fallback if the filter is not applied (e.g., git version differences, flag parsing edge cases). Asserting `output_truncated == false` would make the test fragile against filter variations; the volume guard (`< 2_000 chars`) is the meaningful regression signal.

4. **`chars_threshold_breach` metric field**

   File: `crates/aptu-coder/src/metrics.rs`

   ```rust
   pub struct MetricEvent {
       // ... existing fields ...
       /// True when output_chars exceeds 30,000 -- consistent with MAX_STDOUT_BYTES in #958.
       /// Indicates a large response approaching or past the byte-cap threshold.
       /// Observable via JSONL metrics without waiting for session post-mortem.
       /// Threshold basis: p99.7 of 27,981 observed exec_command calls (2026-05-23 JSONL dataset).
       pub chars_threshold_breach: bool,
   }
   ```

   - Set in the `exec_command` metric emit path as `output_chars > 30_000` (not 20_000 as originally proposed; see Data Basis section above)
   - Default is `false` (not `Option<bool>`) to avoid changing the JSONL schema shape for non-exec tools; new field is additive and backwards-compatible with existing jq one-liners in `AGENTS.md`
   - Add one test: `MetricEvent` with `output_chars = 35_000` has `chars_threshold_breach = true`; with `output_chars = 5_000` has `chars_threshold_breach = false`
   - Document the field in `docs/OBSERVABILITY.md` key schema fields table, including the threshold value and its data basis

### Regression Guard Rationale

The three-layer model after all three issues land:

| Layer | Mechanism | Fires when |
|---|---|---|
| 1. Proactive filter | Built-in/project-local filter table (#959) | Command matches a known-chatty pattern |
| 2. Safety net cap | Byte-level cap (#958) | Any command exceeds per-stream (30k stdout / 10k stderr) or combined (50k) limit |
| 3. Observability | `chars_threshold_breach` metric (this issue) | Any response > 30k chars reaches the LLM (with or without truncation) |

Tests in this issue validate all three layers are wired correctly and that CI catches any regression.

### Integration Notes

- The git repo fixture for the integration smoke test can use `git commit --allow-empty` in a loop with injected file changes; no real remote needed -- use `git pull` against a local second clone acting as the remote
- Existing jq one-liners in `AGENTS.md` use `.tool`, `.duration_ms`, `.result`, `.cache_hit`, `.session_id` -- adding `chars_threshold_breach` is backwards-compatible
- Test file placement: inline `#[cfg(test)]` modules for unit tests; `tests/exec_volume.rs` for the integration smoke test

## Acceptance Criteria

- [ ] `test_line_cap_fires_before_byte_cap` added (the one byte-cap test not covered by #958)
- [ ] All 8 filter unit tests pass (from Strategy 2 above)
- [ ] Integration smoke test `tests/exec_volume.rs` passes: `git pull` on a synthetic 50-commit repo returns < 2,000 chars (no assertion on `output_truncated`)
- [ ] `chars_threshold_breach` field added to `MetricEvent` with threshold 30,000 (not 20,000); set correctly in `exec_command` emit path
- [ ] `chars_threshold_breach` documented in `docs/OBSERVABILITY.md` with threshold value and data basis
- [ ] All existing tests pass: `cargo test`
- [ ] No clippy warnings: `cargo clippy -- -D warnings`
- [ ] Code formatted: `cargo fmt --check`
- [ ] Commit GPG signed and DCO signed-off

## Not In Scope

- Tests for `analyze_*` tools (separate testing surface)
- Load or throughput tests
- Mutation testing

## Additional Context

- Motivation: session `20260523_52` consumed ~70k input tokens for `git pull`; this issue ensures CI would catch a recurrence
- JSONL dataset: 27,981 `exec_command` calls as of 2026-05-23; top-205 calls (output_chars > 20k) account for 17.5% of all output chars
- Threshold change summary: original issue proposed `chars_threshold_breach` at > 20k; revised to > 30k to match `MAX_STDOUT_BYTES` from #958, which was itself sized at p99.67 of the observed distribution. The 20k-30k band (533 calls, 0.41% of total) passes through without truncation; `chars_threshold_breach` at 30k is the early-warning signal for those calls.

### PR DESCRIPTION
**Pull request by clouatre**
## Summary

Adds a regression test suite ensuring `exec_command` output volume never silently regresses. The three-layer guard introduced in #958 and #959 now has full CI coverage. Session `20260523_52` produced 102,582 chars for a single `git pull` call with no automated signal; this PR ensures CI would catch a recurrence.

Closes #960

## Changes

- `crates/aptu-coder/src/metrics.rs`: add `chars_threshold_breach: bool` to `MetricEvent` (threshold 30,000, matching `MAX_STDOUT_BYTES`; fires for the top ~0.33% of `exec_command` calls, p99.7 of 27,981 observed calls). Uses `#[serde(default, skip_serializing_if)]` for backward-compatible JSONL parsing.
- `crates/aptu-coder/src/lib.rs`: emit `chars_threshold_breach = output_chars > 30_000` in the `exec_command` success metric path. All other tool metric emit sites use `..Default::default()` for the new field, eliminating 21 boilerplate repetitions. Adds 9 unit tests: `test_line_cap_fires_before_byte_cap` (gap from #958), 6 filter unit tests, and 2 `MetricEvent` threshold assertions.
- `crates/aptu-coder/tests/exec_volume.rs` (new): integration smoke test -- creates a synthetic 50-commit bare git repo, clones it locally, runs `git pull`, and asserts combined output < 2,000 chars. Does not assert `output_truncated == false` (byte-cap is a valid fallback).
- `docs/OBSERVABILITY.md`: document `chars_threshold_breach` in the key schema fields table with threshold value and data basis.

## Test plan

- [x] 533 tests pass (`cargo test`)
- [x] Linter clean (`cargo clippy -- -D warnings`)
- [x] Formatter clean (`cargo fmt --check`)
- [x] `cargo deny check advisories licenses` clean
- [x] No new external dependencies
- [x] Integration test does not assert `output_truncated == false`
- [x] `chars_threshold_breach` threshold is 30,000 (not 20,000)

### CODE REVIEW

File: crates/aptu-coder/src/lib.rs [hunk 1]
```diff
@@ -891,6 +891,7 @@ impl CodeAnalyzer {
                                 exit_code: None,
                                 timed_out: false,
                                 output_truncated: None,
+                                chars_threshold_breach: false,
```
[the remark this answers is not in the document]
Comment (clouatre) [reply crates/aptu-coder/src/lib.rs]: Done in db1f7f9. Replaced all 21 explicit `chars_threshold_breach: false` sites with `..Default::default()` struct update syntax. The exec_command path retains its explicit `output_chars > 30_000` expression.
